perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_studentIdentifiers.java

```diff
@@ -0,0 +1,115 @@
+package cz.metacentrum.perun.core.impl.modules.attributes;
+
+import cz.metacentrum.perun.core.api.Attribute;
+import cz.metacentrum.perun.core.api.AttributeDefinition;
+import cz.metacentrum.perun.core.api.AttributesManager;
+import cz.metacentrum.perun.core.api.Group;
+import cz.metacentrum.perun.core.api.Status;
+import cz.metacentrum.perun.core.api.User;
+import cz.metacentrum.perun.core.api.exceptions.AttributeNotExistsException;
+import cz.metacentrum.perun.core.api.exceptions.InternalErrorException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeAssignmentException;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+import cz.metacentrum.perun.core.implApi.modules.attributes.UserVirtualAttributesModuleAbstract;
+import cz.metacentrum.perun.core.implApi.modules.attributes.UserVirtualAttributesModuleImplApi;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Get all student identifiers (as list) from any Group where user is valid Member.
+ *
+ * @author Ľuboslav Halama lubo.halama@gmail.com
+ */
+public class urn_perun_user_attribute_def_virt_studentIdentifiers
+	extends UserVirtualAttributesModuleAbstract
+	implements UserVirtualAttributesModuleImplApi {
+
+	private static final String studentIdentifiersValuePrefix = "urn:schac:personalUniqueCode:int:esi:";
+
+	private static final String organizationNamespaceFriendlyName = "organizationNamespace";
+	private static final String organizationScopeFriendlyName = "organizationScope";
+
+	private static final String A_G_D_organizationNamespaceFriendlyName = AttributesManager.NS_GROUP_ATTR_DEF + ":" + organizationNamespaceFriendlyName;
+	private static final String A_G_D_organizationScopeFriendlyName = AttributesManager.NS_GROUP_ATTR_DEF + ":" + organizationScopeFriendlyName;
+
+	private static final String A_U_D_loginNamespaceFriendlyNamePrefix = AttributesManager.NS_USER_ATTR_DEF + ":" + AttributesManager.LOGIN_NAMESPACE + ":";
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_user_attribute_def_virt_studentIdentifiers.class);
+
+	@Override
+	public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDefinition attributeDefinition) {
+		Attribute attribute = new Attribute(attributeDefinition);
+
+		List<Group> userValidGroups = sess.getPerunBl().getMembersManagerBl()
+			.getMembersByUserWithStatus(sess, user, Status.VALID).stream()
+			.flatMap(validMember -> sess.getPerunBl().getGroupsManagerBl().getGroupsWhereMemberIsActive(sess, validMember).stream())
+			.distinct()
```

Comment:
distinct is already used in the method `getAllGroupsWhereMemberIsActive`  so you can omit it here.